CL-2845 - Fix incorrect section title

jamesspeake opened the pull request:
Fixed section1 title to pull input term from phase instead of project for timeline projects

jamesspeake committed:
[CL-2845] Added logic to get the correct input term for a phase

jamesspeake committed:
[CL-2845] Added tests for logic

sebastienhoorens commented on back/app/models/custom_field.rb:
```diff
@@ -182,6 +182,20 @@ def accept(visitor)
+        now = Time.zone.now
+        ideation_types = %w[ideation budgeting]
+        ideation_phases = phases.select { |phase| ideation_types.include? phase.participation_method }
+        current_phase = ideation_phases.each.detect { |phase| phase.start_at <= now && now <= phase.end_at }
```
We have helper methods in the `ParticipationContextService` and `TimelineService` to get the participation context.

jamesspeake replied:
OK, cool someone had already done the hard work! I'll make those changes

jamesspeake replied:
OK, so the timeline service didn't quite have the methods I needed so I've actually added a new method to the timeline service. Better place to put the logic though, and makes it clearer.

sebastienhoorens reviewed (changes requested):
We've had many issues with the input term in the past, and we will keep running into trouble until we introduce the proper abstraction for it (e.g. an `InputTermService`). There are other places in the code where the correct input term must be fetched, so it's important to keep all those places consistent. Out of scope for a quick fix however, but I think we should at least use the helper methods to get the current context. Great test coverage!

jamesspeake committed:
[CL-2845] Refactored logic into timeline service

jamesspeake committed:
[CL-2845] Minor change in logic

sebastienhoorens reviewed (approved):
I wonder why we need to use the first phase as fallback, and not, for example, the last phase that has passed?

Perhaps a small improvement could be to call current_phase from within current_or_first_ideation_phase? Ideally, we should also introduce specs for the new method.

jamesspeake commented:
I guess the main use case for not using current phase is for admin users where a project has finished, so actually I will change to use the last phase - but without introducing any additional logic for whether it is in the past or not

jamesspeake commented:
Reason I didn't call current_phase inside current_or_first_phase is that I wanted to only return ideation phases first

jamesspeake committed:
[CL-2845] Updated to use the last ideation phase, not the first

jamesspeake committed:
[CL-2845] Added tests for the timeline service

jamesspeake commented:
@sebastienhoorens I know you've already approved these, but as I've changed the logic to last phase and added some tests for the timeline service, it would be good for you to take one last quick look

sebastienhoorens reviewed (changes requested):
I don't think it's correct to use the input term of the last phase as fallback. The project schema endpoint uses the project input term as fallback instead, could we do the same so that the input term remains consistent everywhere? See project_custom_fields_controller:22.

jamesspeake commented:
For me I think this would be the most incorrect option as for a timeline project you cannot edit the default input term. It does fall back to that but only if there are no ideation phases. Also Amanda has implemented the same logic on the front-end. If it's OK with you I will raise another ticket to look at making consistent later.

sebastienhoorens reviewed (approved):
Alright, would be good to review this indeed, and not duplicate this logic (also not in the frontend).
